tsl/src/chunk.c

```diff
@@ -1180,3 +1186,538 @@ chunk_merge_chunks(PG_FUNCTION_ARGS)
+						HeapTupleHeaderGetUpdateXid(tuple->t_data)))
+					elog(WARNING,
+						 "concurrent delete in progress within table \"%s\"",
+						 RelationGetRelationName(splitrel));
+				/* treat as recently dead */
+				tups_recently_dead += 1;
+				isdead = false;
+				break;
+			default:
+				elog(ERROR, "unexpected HeapTupleSatisfiesVacuum result");
+				isdead = false; /* keep compiler quiet */
+				break;
+		}
+
+		LockBuffer(buf, BUFFER_LOCK_UNLOCK);
+
+		/* Find the partition/chunk to insert the tuple into */
+		Datum value = slot_getattr(srcslot, splitdim_attnum, &isnull);
+		Assert(!isnull);
+		int64 point = ts_time_value_to_internal(value, splitdim_type);
+
+		/*
+		 * Route to partition based on new boundaries. Only 2-way split is
+		 * supported now, so routing is easy. An N-way split requires, e.g.,
+		 * binary search.
+		 */
+		if (point < split_point)
+			rsi = splitinfos[0];
+		else
+			rsi = splitinfos[1];
+
+		if (isdead)
+		{
+			tups_vacuumed += 1;
+			/* heap rewrite module still needs to see it... */
+			if (rewrite_heap_dead_tuple(rsi->rwstate, tuple))
+			{
+				/* A previous recently-dead tuple is now known dead */
+				tups_vacuumed += 1;
+				tups_recently_dead -= 1;
+			}
+			continue;
+		}
+
+		num_tuples++;
+		reform_and_rewrite_tuple(tuple, splitrel, rsi);
+	}
+
+	MemoryContextSwitchTo(oldcxt);
+
+	const char *nspname = get_namespace_name(RelationGetNamespace(splitrel));
+
+	ereport(elevel,
+			(errmsg("\"%s.%s\": found %.0f removable, %.0f nonremovable row versions",
+					nspname,
+					RelationGetRelationName(splitrel),
+					tups_vacuumed,
+					num_tuples),
+			 errdetail("%.0f dead row versions cannot be removed yet.", tups_recently_dead)));
+
+	table_endscan(scan);
+	ExecDropSingleTupleTableSlot(srcslot);
+	FreeExecutorState(estate);
+
+	ts_cache_release(hcache);
+	table_close(splitrel, NoLock);
+
+	/* Cleanup split rel infos and reindex new chunks */
+	for (int i = 0; i < 2; i++)
+	{
+		ReindexParams reindex_params = { 0 };
+		int reindex_flags = REINDEX_REL_SUPPRESS_INDEX_USE;
+		Oid chunkrelid = splitinfos[i]->rel->rd_id;
+
+		if (relpersistence == RELPERSISTENCE_UNLOGGED)
+			reindex_flags |= REINDEX_REL_FORCE_INDEXES_UNLOGGED;
+		else if (relpersistence == RELPERSISTENCE_PERMANENT)
+			reindex_flags |= REINDEX_REL_FORCE_INDEXES_PERMANENT;
+
+		relation_split_info_free(splitinfos[i], ti_options);
+
+		/* Only reindex new chunks. Existing chunk will have be reindexed
+		 * during the heap swap below. */
+		if (i > 0)
+			reindex_relation_compat(NULL, chunkrelid, reindex_flags, &reindex_params);
+	}
+
+	/* Finally, swap the heap of the chunk that we split so that it only
+	 * contains the tuples for its new partition boundaries. */
+	finish_heap_swap(relid,
+					 new_heap_relid,
+					 false, /* system catalog */
+					 false /* swap toast by content */,
+					 true, /* check constraints */
+					 true, /* internal? */
+					 cutoffs.FreezeLimit,
+					 cutoffs.MultiXactCutoff,
+					 relpersistence);
```

Comment:
This is safe since you do not release the AccessExclusiveLock when you close the table above, but you might want to add that to the comment.

If somebody decides it is a good idea to release the access exclusive lock before this, there is a window where tuples can be inserted, and then swapped out (hence lost) because you switch the underlying storage.

Reply:
Added comment.